### ISSUE TITLE
[Semantic Configuration] A config TreeBuilder

### ISSUE THREAD

**Issue by Djeg** [KnpLabs/KnpRadBundle#88]
I have noticed that actually the RadBundle does not permit to define a correct config TreeBuilder. It would be interesting to give this kind of functionality. Maybe simplify it a little bit.

I am thinking about how to improve the actual AppBundle and add two methods like : 

``` php
public function buildTree(NodeParentInterface $rootNode);

public function buildContainer(array $config, ContainerBuilder $container)
```

The `buildContainer` method corespond to the standard `load` method of an bundle Extension object. Except that the given `$config` in this new method is already parsed.

What did you think about this idea ?

**Comment by docteurklein** [comment]
No, you still can define a good old `DependencyInjection/AppExtension` class that would be loaded [instead of the default one](https://github.com/KnpLabs/KnpRadBundle/blob/develop/AppBundle/Bundle.php#L26).

Then, in this custom extension class, instanciate your Configuration class where you configure your TreeBuilder.

The process I described is just good old symfony stuff.

But there would indeed be a possiblity to make stuff easier by defining a `getConfigTreeBuilder` method in the App bundle, and then call:

```
        $config = (new Processor)->process($this->getConfigTreeBuilder()->buildTree(), $configs);
```

 in the default knpRad extension.

**Comment by docteurklein** [comment]
Then, your bundle class may do too much stuff, for what it should not be aware of.

**Comment by docteurklein** [comment]
let me do a POC / **EDIT**: I don't see the use case

**Comment by docteurklein** [comment]
BTW, what's the purpose to create a semantic config (using TreeBuilder) if you do not override the Extension class?
You have no other access to it but in the extension class.

**Comment by docteurklein** [comment]
if you want to add DI parameters, just use parameters in app/config:

``` yaml

parameters:
    app.config:
        enabled: true
        test: false

```

relates to https://github.com/KnpLabs/KnpRadBundle/pull/77

**Comment by Djeg** [comment]
Yeah sorry. We can defined a good old `DependencyInjection\AppExtension` this is corect but i think it's interesting to simplify the different steps. Anyway, i agree with the "Too much stuff". 

My first intention has been to create an `AbstractConfigurationExtension` object that would contains the `buildTree` and `buildContainer` shortcuts. The user have to respect this abstract class into the `DependencyInjection/ConfigurationExtension` (maybe ...) and us, we load this stuff in the current KnpRadExtension.

Concretly my intention is to provide a simple way to add TreeBuilding configuration stuff for application that need maybe a lot of configuration. the `config:dump-reference` command is very interesting and powerfull ! And configuration still always well formated (and documented). I think it's a realy good practice that need to be simplify in the RadBundle ? Don't you ?

**Comment by docteurklein** [comment]
ok to have a simple approach for defining semantig config and be able to dump it from console! :)

But the problem is, what to do then with this config ? put it rawly in the DI parameters?

**Comment by Djeg** [comment]
By default the method implemented in the base rad bundle should work this way.

**Comment by Djeg** [comment]
And anyway, the bundle class is empty, so using it to define "configuration of the bundle" seems interesting i guess ?

**Comment by docteurklein** [comment]
yes indeed!

**Comment by docteurklein** [comment]
Another point @davidjegat: by making the `RadExtension` inherit from `ConfigurableExtension`, we should be able to just add a class named `DependencyInjection/Configuration` and it would be automatically loaded. I'm testing right now.

**Comment by docteurklein** [comment]
hey, here is an example of how to do it without modifying anything:

``` php

<?php

namespace App;

use Knp\RadBundle\AppBundle\Bundle;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Extension\ExtensionInterface;
use Symfony\Component\Config\Definition\ConfigurationInterface;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;

class App extends Bundle implements ExtensionInterface, ConfigurationInterface
{
    public function getConfigTreeBuilder()
    {
        $tree = new TreeBuilder;
        $rootNode = $tree
            ->root('app')
            ->children()
                ->scalarNode('test')
            ->end()
        ->end();

        return $tree;
    }

    public function load(array $config, ContainerBuilder $container)
    {
        $extension = new ContainerExtension($this->getPath());
        $extension->load($config, $container);

        // do more stuff
    }

    public function getAlias()
    {
        return 'app';
    }

    public function getContainerExtension()
    {
        return $this->extension = $this;
    }

    public function getConfiguration(array $configs, ContainerBuilder $container)
    {
        return $this;
    }

    public function getXsdValidationBasePath()
    {
        return false;
    }

    public function getNamespace()
    {
        return 'http://example.org/schema/dic/'.$this->getAlias();
    }
}


```

**Comment by docteurklein** [comment]
another possiblity would be to make the App bundle only implements `ConfigurationInterface`.

Then modify the `getConfiguration` method to:

``` php
    public function getConfiguration()
    {
        if (null !== $config = parent::getConfiguration()) {
            return $config;
        }

        if ($this->bundle instanceof ConfigurationInterface) {
            return $this->bundle;
        }
    }
```

Then, you just have to implements getConfigTreeBuilder method, but again, then what to do with semantic $config.

**Comment by lsmith77** [comment]
did you mean to close the ticket?

**Comment by docteurklein** [comment]
yes. It has been implemented by #89.
